tgui/packages/tgui/interfaces/PlaneMasterDebug/index.tsx

```diff
@@ -0,0 +1,640 @@
+    .concat(plane.outgoing_relays)
+    .map((connection: Filter | Relay) => connection.target);
+  const allElems: Plane[] = [];
+  let foundChild = false;
+  while (checkDepth <= 2) {
+    let newCheck: Plane[] = [];
+    for (let i = 0; i < toCheck.length; i++) {
+      const checkElem = toCheck[i];
+      if (checkElem === undefined) {
+        continue;
+      }
+
+      if (allElems.includes(checkElem)) {
+        foundChild = true;
+        break;
+      }
+
+      allElems.push(checkElem);
+      newCheck = newCheck.concat(
+        ((checkElem as Plane).outgoing_filters as (Filter | Relay)[])
+          .concat(checkElem.outgoing_relays)
+          .map((connection: Filter | Relay) => connection.target)
+          .filter(isDefined),
+      );
+    }
+
+    if (foundChild) {
+      break;
+    }
+
+    toCheck = newCheck;
+    checkDepth++;
+  }
+
+  // If our plane has at least 2 children over 2 degrees of separation away
+  // from each other, give them a bump in depth to split them from other
+  // parents of those planes visually a bit
+  if (checkDepth) {
+    depth += 1;
+  }
+
+  plane.depth = Math.max(depth, plane.depth);
+  for (let i = 0; i < plane.parents.length; i++) {
+    evaluatePlaneDepth(plane.parents[i], depth + 1);
+  }
+}
+
+// Stolen wholesale from fontcode
+function textWidth(text: string, font: string, fontsize: number) {
+  // default font height is 12 in tgui
+  font = `${fontsize}px ${font}`;
+  const c = document.createElement('canvas');
+  const ctx = c.getContext('2d') as CanvasRenderingContext2D;
+  ctx.font = font;
+  return ctx.measureText(text).width;
+}
+
+function getPlaneHeight(plane: Plane) {
+  return (
+    45 +
+    19 *
+      Math.max(
+        plane.incoming_filters.length + plane.incoming_relays.length,
+        plane.outgoing_filters.length + plane.outgoing_relays.length + 1,
+      ) +
+    15
+  );
+}
+
+function getDesiredPlanePosition(
+  plane: Plane,
+  curStack: number,
+  tallestStack: number,
+) {
+  const dependents: Plane[] = (
+    curStack > tallestStack
+      ? (plane.outgoing_filters as (Filter | Relay)[]).concat(
+          plane.outgoing_relays,
+        )
+      : (plane.incoming_filters as (Filter | Relay)[]).concat(
+          plane.incoming_relays,
+        )
+  )
+    .map((connection: Filter | Relay) =>
+      curStack > tallestStack ? connection.target : connection.source,
+    )
+    .filter(isDefined);
+
+  const avgY =
+    dependents
+      .map((x) => x.position.y + getPlaneHeight(x) / 2)
+      .reduce((a, b) => a + b, 0) / dependents.length;
+
+  return avgY - getPlaneHeight(plane) / 2;
+}
+
+export function PlaneMasterDebug() {
+  const { data, act } = useBackend<PlaneDebugData>();
+  const {
+    mob_name,
+    planes,
+    tracking_active,
+    mob_ref,
+    our_ref,
+    enable_group_view,
+    our_group,
+    present_groups,
+  } = data;
+  const connectionDom = useRef<PlaneConnectorsMap>({});
+
+  const planesProcessed = useMemo(() => {
```

Comment:
This should be made into an external function or custom hook. You could still use usememo for the function return